replica/distributed_loader.cc

```diff
@@ -465,22 +462,24 @@ distributed_loader::process_upload_dir(distributed<replica::database>& db, distr
         };
         process_sstable_dir(directory, flags).get();
 
-        auto generation = highest_generation_seen(directory).get0();
-        auto shard_generation_base = generation.value_or(replica::table::make_new_generation()).value() / smp::count + 1;
-
-        // We still want to do our best to keep the generation numbers shard-friendly.
-        // Each destination shard will manage its own generation counter.
-        std::vector<std::atomic<sstables::generation_type::int_t>> shard_gen(smp::count);
-        for (shard_id s = 0; s < smp::count; ++s) {
-            shard_gen[s].store(shard_generation_base * smp::count + s, std::memory_order_relaxed);
-        }
-
-        // we need generation calculated by instance of cf at requested shard
-        auto new_generation_for_shard = [&] (shard_id shard) {
-            auto gen = shard_gen[shard].fetch_add(smp::count, std::memory_order_relaxed);
-            return sstables::generation_type(gen);
+        sharded<sstables::sstable_generation_generator> sharded_gen;
+        auto highest_generation = highest_generation_seen(directory).get0().value_or(
+            sstables::generation_type{0});
+        sharded_gen.start(
+            sharded_parameter([highest_generation] {
+                // get the base generation so we can increment it by smp::count without
+                // conflicting with other shards
+                return highest_generation.value() / smp::count * smp::count + this_shard_id();
+            })).get();
```

Comment:
I think this logic should live in sstables::sstable_generation_generator. It's irrelevant to UUID.

Also note that value / smp::count * smp::count + this_shard_id() will align down.

we need to add smp::count to value, before aligning it down.

Reply:
please see the implementation of the generator, we always add smp::count before returning the next id. so no need to add smp::count.

Reply:
Alright. But please let's move all this arithmetic there. It's generation type specific. An implementation detail that the user doesn't have to be aware of.